File: app/src/main/java/com/javernaut/whatthecodec/VideoFileConfig.java
package com.javernaut.whatthecodec;

import android.graphics.Bitmap;
import android.os.ParcelFileDescriptor;

public class VideoFileConfig {

    // The field is handled by the native code
    private long nativePointer;

    private VideoFileConfig(int fileDescriptor) {
        nativeNew(fileDescriptor);
    }

    public native String getFileFormat();

    public native String getCodecName();

    public native int getWidth();

    public native int getHeight();

    public native void release();

    public native void fillWithPreview(Bitmap bitmap);

    public static VideoFileConfig create(ParcelFileDescriptor descriptor) {
        VideoFileConfig result = new VideoFileConfig(descriptor.detachFd());
        if (result.nativePointer == -1) {
            return null;
        }
        return result;
    }

    private native void nativeNew(int fileDescriptor);

    static {
        System.loadLibrary("avformat");
        System.loadLibrary("avcodec");
        System.loadLibrary("avutil");
        System.loadLibrary("swscale");
        System.loadLibrary("video-config");
    }
}

File: app/src/main/java/com/javernaut/whatthecodec/VideoFileConfig.kt
package com.javernaut.whatthecodec

import android.graphics.Bitmap
import android.os.ParcelFileDescriptor

class VideoFileConfig(val fullFeatured: Boolean) {

    private constructor(fileDescriptor: Int) : this(false) {
        nativeNewFD(fileDescriptor)
    }

    private constructor(filePath: String) : this(true) {
        nativeNewPath(filePath)
    }

    // The field is handled by the native code
    private val nativePointer: Long = 0

    val fileFormat: String
        external get

    val codecName: String
        external get

    val width: Int
        external get

    val height: Int
        external get

    external fun release()

    external fun fillWithPreview(bitmap: Array<Bitmap>): Boolean

    private external fun nativeNewFD(fileDescriptor: Int)

    private external fun nativeNewPath(filePath: String)

    companion object {

        fun create(filePath: String) = returnIfValid(VideoFileConfig(filePath))

        fun create(descriptor: ParcelFileDescriptor) = returnIfValid(VideoFileConfig(descriptor.detachFd()))

        private fun returnIfValid(config: VideoFileConfig) =
                if (config.nativePointer == -1L) {
                    null
                } else config

        init {
            System.loadLibrary("avutil")
            System.loadLibrary("avcodec")
            System.loadLibrary("avformat")
            System.loadLibrary("swscale")
            System.loadLibrary("video-config")
        }
    }
}
